### ISSUE TITLE
Take photo using Webcam

### ISSUE THREAD

**Issue by brunobsouzac** [telegramdesktop/tdesktop#781]
![2](https://cloud.githubusercontent.com/assets/12419098/7947907/77646528-0956-11e5-9ea3-e6a48642a2cd.png)

Sometimes I want to take a photo by Telegram Desktop, but there is only the option to send a photo. you can use the webcam to take a picture would be ideal. Something more elaborate and will be asked here a while is the links and video calling. Could release the webcam for video calling as it is more interesting in the desktop Smartphones.

Thank you for all devs!

**Comment by auchri** [comment]
#414

**Comment by brunobsouzac** [comment]
@auchri thanks

**Comment by bil-elmoussaoui** [comment]
If it's a duplicate, the best is to close this :P

**Comment by auchri** [comment]
Yes. Or close the old one :stuck_out_tongue_winking_eye:

**Comment by brunobsouzac** [comment]
It is a duplicate, true. But more recent and images, it would be an attractive best perhaps. It's a question that few put only makes sense since the client is for desktop. What do you think?

**Comment by brunobsouzac** [comment]
I mentioned in #414 

It's a safety issue, but it makes sense to have this option as it is focused on the desktop. A user get the smartphone to take pictures becomes much more complicated and we should focus on usability for both desktop and mobile. The telegram it is in need of a reformed, be simple, safe and objective.

Take a picture using the webcam also opens a possibility being requested by users is video calls and connections as well as Skype, but open, simple and secure. Of course the community will decide this with time.

**Comment by brunobsouzac** [comment]
what do you think of this idea ?? looks at the end of the reviews (I decided to make a mockup using the paint) https://github.com/DrKLO/Telegram/issues/1147#issuecomment-111154097 New UX/UI Design

**Comment by ppparadox** [comment]
+1

**Comment by corb06** [comment]
Is this feature planned for future releases?

**Comment by brunobsouzac** [comment]
@corb06 I hope so. And that puts the shortcut ctrl + space to send voice messages instead of pressing and holding the audio button

**Comment by Dea1993** [comment]
+1 add function to take pictures from webcam, like on cutegram, is very useful.

**Comment by andrewxtsang** [comment]
+1 for screenshot capture
Line has this function in desktop
Also, should rename the title as #838 is closed

**Comment by corb06** [comment]
Yeah as of now all issues regarding selectable screen area for capture and send has been closed.

I will have to reopen another one.

**Comment by wildeyes** [comment]
+1 this has to happen already.

**Comment by Aokromes** [comment]
+1 i will like to use the integrated camera of my tablet to take photos :)

**Comment by tommybrc** [comment]
This is almost a must already. I see no reason why the mobile version sends photos with any problem, while the desktop one has security issues with that! ¿Why? A lot less secure is to be able to send files!!

Having video calls would be great, but I think it should be required as a different issue, if it hasn't been already.

**Comment by auchri** [comment]
>  I see no reason why the mobile version sends photos with any problem, while the desktop one has security issues with that!

Who says something about security issues?

**Comment by Aokromes** [comment]
https://github.com/telegramdesktop/tdesktop/issues/781#issuecomment-108109171 did :)

**Comment by auchri** [comment]
But he is not a member of the telegram team, so it is not necessarily true what he says :wink:

**Comment by tommybrc** [comment]
- brunobsouzac commented on 2 Jun 2015: "It's a safety issue,..."
- ryanralph commented on 20 Jan 2015 on issue #414: "I disagree. As a security conscious person I would actually prefer to isolate...."

And a few more I don´t want to search now.

**Comment by mirh** [comment]
Who are still nobody. So please, leave the dev wonder how to implement straightforwardly something so code-burdening.

**Comment by tommybrc** [comment]
I leave them all do whatever they have to do. And I hope they are able to implement something asap! This would be a very nice functionality to have! Best regards!

**Comment by DiThi** [comment]
Tdesktop has ffmpeg as dependency (used to e.g. converts gifs to mp4s). It shouldn't be difficult to use it for webcam input, as It includes support for directshow and v4l2.

https://trac.ffmpeg.org/wiki/DirectShow

**Comment by ADTC** [comment]
Oh, why not? I'm actually quite surprised that you can use the microphone to record sounds but you cannot use the webcam to take pictures and record videos.

I also like the ability to take a screenshot. Perhaps the menu can be expanded, like how you see it in mobile:

- File
- Camera (use current Photo icon)
- Photo/Video (use a photo frame icon with a small video icon offset)
- Screenshot (use a monitor with dragged box icon)

**Camera** should show a webcam window in the middle of Telegram window, and activate the webcam of the computer. Press to take a photo, or hold to record a video.

**Photo/Video** should open the dialog to select files. The filter should be 'Images and Video Files'. Videos could be compressed as well before sending, rather than sending huge files.

**Screenshot** should minimize Telegram window and activate an overlay where the user can drag a section of the screen to take a screenshot. If user held Ctrl while clicking _Screenshot_, or right-clicked _Screenshot_, Telegram window should not minimize but stay on screen (to be included in screenshot if user desired). _There should be a guide for first-time activation of this feature to tell the user how to use it (including the Ctrl/right-click modification)._

**_PS:_** I don't really understand the security argument against webcam access, since messengers like Skype, Yahoo Messenger, etc are normally expected to be able to access webcam as part of their normal messaging experience, and one would expect the same with Telegram as a messenger. In fact, Telegram on mobile has full access to both cameras on a phone - including a recently introduced "previewing shortcut" to the Telegram's internal camera function. If a user is really paranoid about their faces being recorded in secret, they can disable Telegram's access to the webcam hardware, they can disable the webcam hardware or they can even physically block the webcam's aperture from capturing any light (there are those sliding window thingies to stick on a laptop webcam aperture you can find nowadays). **But to suggest that you inconvenience everyone else with having to go through a whole lot of hoops to achieve such a simple functionality, JUST so that your paranoia is well served and taken care of is downright ridiculous.**

**Comment by umagon** [comment]
Will this functionality be made?

**Comment by Aokromes** [comment]
Few days ago the battery of my mobile phone was exhausted, no gps, so i had to take a photo with windows camera application and send it with this way we can save time :)

**Comment by Aokromes** [comment]
Open source multiplataform library for this https://opencv.org/

**Comment by uzielweb** [comment]
Duplicated or not, secure or not. Telegram needs this two:
- video and photo via webcam in desktop app
- create a videocall in all apps

**Comment by Kyrluckechuck** [comment]
This is actually quite insane that it's not implemented yet!

Facebook for the web can even do it, let alone any other standalone messaging app which you install. 

Can we either get the go-ahead to implement this or a solid reason why not?

It _seems_ to just be getting ignored because someone doesn't agree with the rest of us about this, but not actually providing any concrete reasoning...

**Comment by Kyrluckechuck** [comment]
@john-preston Thank you for the response, that helps add some reasoning to it.

So I assume that if someone else were to create a pull request about this that didn't force the libpulse dependency it would likely get the A-OK?

Is the hesitation with requiring libpulse for building or for runtime usage on unix-like systems?

**Comment by john-preston** [comment]
@Kyrluckechuck The problem with libpulse is that if it is a static dependency and someone without libpulse will autoupdate tdesktop his version will be replaced with one that won't launch on his system anymore.

If someone makes a PR that adds a way to take a photo from a webcam and it won't add new static dependencies, this will be cool. It even can work not in all cases (not on all platforms / versions / DEs), but it should work at least on modern Windows / macOS versions and it should provide a way to check if it will work or not (so that UI elements could be hidden and so on), without bad user experience (like there is always a button, but no one knows will it work or not if clicked).

**Comment by Kyrluckechuck** [comment]
Absolutely, that makes total sense. 

Thanks for the information, hopefully this will help move a PR along!

**Comment by jmallach** [comment]
Hey @john-preston, I don't understand: what happens if a new version of tdesktop starts depending on libpulse0? What if qtcore starts depending on something new? At some point you will need to add or remove or update dependencies, right? Let's see if I get it: this is in the case of people autoupdating the tdesktop binary. What happens if someone installs tdesktop without installing all the Qt libraries, for example? Isn't it the same case?

**Comment by john-preston** [comment]
@jmallach Qt is linked statically. I don't think QtCore / QtGui / QtWidgets / QtNetwork will start depending on non-statically-linkable dynamic library.

**Comment by Johnnynator** [comment]
>  What happens if someone installs tdesktop without installing all the Qt libraries, for example? Isn't it the same case?

Qt is statically linked and therefore not needed on the system. It links only against these.
```
librt.so.1
libdrm.so.2
libz.so.1
libICE.so.6
libSM.so.6
libfontconfig.so.1
libfreetype.so.6
libEGL.so.1
libGL.so.1
libxcb.so.1
libX11.so.6
libX11-xcb.so.1
libdl.so.2
libglib-2.0.so.0
libpthread.so.0
libm.so.6
libgcc
libc.so.6
ld-linux-x86-64.so.2
```

**Comment by jmallach** [comment]
Ah, I see. So how do you deal with a system not having libglib or libfontconfig installed? Or do you consider all systems running a desktop environment will surely have those two? (I ask because I thought libpulse was pretty much the norm these days, too)

**Comment by corb06** [comment]
5 years and still not implemented. Whatsapp have had this feature for many many years

**Comment by Aokromes** [comment]
> 5 years and still not implemented. Whatsapp have had this feature for many many years

whatsapp "desktop" have support of take photo? :)

**Comment by corb06** [comment]
> > 5 years and still not implemented. Whatsapp have had this feature for many many years
> 
> whatsapp "desktop" have support of take photo? :)

Yepp! 2 clicks and photo taken. 
![Untitled](https://user-images.githubusercontent.com/2758407/76106906-16e67c80-5fd8-11ea-907f-22b808632e28.jpg)

**Comment by purr** [comment]
This should be at the top of the list for addition if you want to be a good communication tool, especially in the 21st century every communication app has video call/screenshare expect this one!! makes it hard to switch when you like to video call your friends

**Comment by frzifus** [comment]
It would already be a great feature. Especially because people under Sailfish, UBPorts, fedora or Mobian use telegram desktop on their phones 😉 there a camera device is nothing more than a webcam.
in terms of security, access can still be limited via bw. At least on Linux.

**Comment by azat11** [comment]
Yeah, telegram is not really quick on adding new features, that's for sure. But then again their team is not as big as Facebook owned WhatsApp's, I think...

**Comment by uzielweb** [comment]
Please! We need to do all actions in all paltforms to this important request.

**Comment by barfin** [comment]
this would be a great feature

**Comment by WrongBit** [comment]
It's 7th March and... still nobody is assigned? Guys, what the "serious rubbish" you're busy with that this high demand feature still hangs w/o attention?
You made chat client and obviously "take pic from camera" is quite needed feature. I request to pay MORE attention to this.

**Comment by jyxjjj** [comment]
WinKey+Shift+S can get screenshot or use PrintScreen and some others in Windows.
But of cause we need some for Linux.
And yes, we request a feature to use camera.

**Comment by barfin** [comment]
> 
> 
> It's 7th March and... still nobody is assigned? Guys, what the "serious rubbish" you're busy with that this high demand feature still hangs w/o attention?
> You made chat client and obviously "take pic from camera" is quite needed feature. I request to pay MORE attention to this.

I totally agree

**Comment by DrHeidi** [comment]
Yes please... I would really like to be able to take a picture with my laptop webcam, similar to my iPad and iPhone...

**Comment by NightMachinary** [comment]
> Yeah, telegram is not really quick on adding new features, that's for sure. But then again their team is not as big as Facebook owned WhatsApp's, I think...

Let's be real here. WhatsApp still does not have even an iPad app. The only "innovation" I have seen from them in recent memory is that they have implemented a progressive get-shittier-or-agree-to-new-abuse-terms system, which I haven't tried, but I think their big team is just enough to get right.

**Comment by macydnah** [comment]
> This should be at the top of the list for addition if you want to be a good communication tool, especially in the 21st century every communication app has video call/screenshare expect this one!! makes it hard to switch when you like to video call your friends

Totally agree, this should be top on additiomn list. Please add it, would be an awesome feature

**Comment by Aokromes** [comment]
https://github.com/telegramdesktop/tdesktop/commit/3cb595c3c9add29d0f2fa6cfa760de1df9594fa7

maybe this code can be updated to allow to take photos with any place?

**Comment by maximecharriere** [comment]
+1

**Comment by AmadeusK525** [comment]
> [3cb595c](https://github.com/telegramdesktop/tdesktop/commit/3cb595c3c9add29d0f2fa6cfa760de1df9594fa7)
> 
> maybe this code can be updated to allow to take photos with any place?

Huh, that looks pretty promising. Now that I think about it, how are video calls handled in the app? Since they already have a way to connect to the camera, I would assume that @john-preston's worries about the dependencies in 2018 don't apply anymore. Anyhow, I'm sure the profile picture code could be abstracted into a common camera access for both video messages and pictures (although the video messages are probably harder)

**Comment by john-preston** [comment]
@raminrez @ogajvoronsky This interface is not from Telegram Desktop. I believe you're using native Telegram app: https://github.com/overtake/TelegramSwift

**Comment by raminrez** [comment]
@john-preston  I installed the app from the app store. even tried to download from the official website and use it. steel has the problem!


It shows some photos taken by the telegram app, but I need to remove the recent ones! 
the better idea is to remove the native behaviour of taking photos on the mac.

**Comment by john-preston** [comment]
@raminrez This is another app. If you want Telegram Desktop you can try it from https://desktop.telegram.org

**Comment by LexaGV** [comment]
7 years "telegram team" discusses obviously useful feature. It's a shame! For 7 years MS made a whole operating system! And only team clowns discuss "do we need function to take a photo?....". Unbelievable stupidity and laziness!

**Comment by bluemooduser** [comment]
Really? 7 years ago someone asks for something so simple and it has not been implemented yet? 
What a pity!

**Comment by ilya-fedin** [comment]
@bluemooduser a lot of words are applicable to the feature, but it's definitely not simple

**Comment by john-preston** [comment]
Now it could be done easily, but there is no UI place to put it to 🙈  Maybe an attach icon context menu? 🤔

**Comment by LexaGV** [comment]
Doh... 7 years more to discuss where to place icon? Or JUST MAKE IT?

**Comment by ilya-fedin** [comment]
@LexaGV if you think it's that simple then maybe you will make it?

**Comment by LexaGV** [comment]
@ilya-fedin : Sure! Just make me paid employee. :) I made this feature in 2 hours (in my WPF app). You make it 7 years. Don't you feel blush??

**Comment by ilya-fedin** [comment]
@LexaGV you're forgetting the fact tdesktop written in C++ and is a cross-platform app. You won't have the convenient APIs WPF provide you if you worked on tdesktop.

**Comment by LexaGV** [comment]
Well, Telegram team made two big mistakes at very start, now is too late to fix.
1. C++ is the latest you should select for development. C# outperforms C++ in the long term. And obviously messenger is not the "CPU intensive" app.
2. There is no real demand for crossplatform - it's just wet dreams of unexperienced developers. You need literally two apps - Windows/WPF and Android. Both could be native and don't suffer from ugly "browser" underneath.

Not sure can it help, but I used old "DirectShow" library for .NET: https://directshownet.sourceforge.net/ . What I'm sure, C++ must have API for DirectShow and you can easy implement grabbing camera frames. Trust me, this feature worth to implement!

**Comment by ilya-fedin** [comment]
DirectShow sadly is only for Windows, while features are usually implemnted simultaneously for all platforms, so it needs researching all the platforms at once. But I guess according to @john-preston it's already implemented, apparently using webrtc APIs.

> Trust me, this feature worth to implement!

There are features to implement every month. New features are top priority. Given the amount of work and that @john-preston usually have no spare time after implementing new features, we're getting something like 7 years to implement a feature of such size.

**Comment by LexaGV** [comment]
@ilya-fedin "DirectShow sadly is only for Windows, while features are usually implemnted simultaneously for all platforms" - your "multiplatform consistency" already suffers from the fact Android TG **has** feature to capture camera! Only Windows version lack of it. Second, nobody forces you to use ONE library for all platforms! (obviously they are different) On Windows it's DirectShow, on Linux is smth like FrameBuffer... the only thing should be in common is UI.

**Comment by ilya-fedin** [comment]
@LexaGV the point was implementing 3 APIs means 3x times more work, not that anyone forces to one API. On Linux it's not framebuffer, but V4L2/PipeWire/libcamera (yeah, Linux multiple APIs even for such a thing). And something for macOS.

**Comment by LexaGV** [comment]
@ilya-fedin Whatever you name, TG for Windows does NOT have camera capture, while Android version has. It's already inconsistency of your "artificial policy" and lack of functionality. Make camera capture on Windows is not so huge job and sure, more important than idiotic animations on 1/2 of screen.

I recommend TG team to publish voting for features - it gives much better picture what people wants (instead of sitting 3 guys in a room and fantasy about our wishes).

**Comment by ilya-fedin** [comment]
>  sure, more important than idiotic animations on 1/2 of screen.

That's sadly not for you to decide

> I recommend TG team to publish voting for features - it gives much better picture what people wants (instead of sitting 3 guys in a room and fantasy about our wishes).

(Un)fortunately no one who decides reads github

> Make camera capture on Windows is not so huge job

It's a huge job. At least it was until @john-preston's message.

> Whatever you name, TG for Windows does NOT have camera capture, while Android version has. It's already inconsistency of your "artificial policy" and lack of functionality.

Whatever you say, this won't end in an implementation. Unless you want to really help.
